src/compute-client/src/protocol/history.rs

```diff
@@ -180,49 +179,92 @@ where
-        command_count += live_cancels.len();
-        if !final_configuration.all_unset() {
-            command_count += 1;
-        }
-
-        self.metrics.reset();
+        // For commands that batch multiple entries, we count the number of entries instead of the
+        // number of commands. This is to ensure that history reduction is not invoked too often.
+        self.reduced_count = 0;
 
         // Reconstitute the commands as a compact history.
+        // When we update `self.metrics`, we need to be careful to not transiently report incorrect
+        // counts, as they would be ovservable by other threads.
+
+        let metric = self.metrics.command_counts.create_timely.borrow();
         if let Some(create_timely_command) = create_timely_command {
-            self.push_inner(create_timely_command);
+            metric.set(1);
+            self.reduced_count += 1;
+            self.commands.push(create_timely_command);
+        } else {
+            metric.set(0);
         }
+
+        let metric = self.metrics.command_counts.create_instance.borrow();
         if let Some(create_inst_command) = create_inst_command {
-            self.push_inner(create_inst_command);
+            metric.set(1);
+            self.reduced_count += 1;
+            self.commands.push(create_inst_command);
+        } else {
+            metric.set(0);
         }
+
+        let metric = self.metrics.command_counts.update_configuration.borrow();
         if !final_configuration.all_unset() {
-            self.push_inner(ComputeCommand::UpdateConfiguration(final_configuration));
+            metric.set(1);
+            self.reduced_count += 1;
+            self.commands
+                .push(ComputeCommand::UpdateConfiguration(final_configuration));
+        } else {
+            metric.set(0);
         }
+
+        let dataflow_count = u64::cast_from(live_dataflows.len());
+        self.metrics.dataflow_count.borrow().set(dataflow_count);
+        let metric = self.metrics.command_counts.create_dataflows.borrow();
         if !live_dataflows.is_empty() {
-            self.push_inner(ComputeCommand::CreateDataflows(live_dataflows));
+            metric.set(1);
+            self.reduced_count += live_dataflows.len();
+            self.commands
+                .push(ComputeCommand::CreateDataflows(live_dataflows));
+        } else {
+            metric.set(0);
         }
+
+        let peek_count = u64::cast_from(live_peeks.len());
+        self.metrics.command_counts.peek.borrow().set(peek_count);
         for peek in live_peeks {
-            self.push_inner(ComputeCommand::Peek(peek));
+            self.reduced_count += 1;
+            self.commands.push(ComputeCommand::Peek(peek));
         }
+
+        let metric = self.metrics.command_counts.cancel_peeks.borrow();
         if !live_cancels.is_empty() {
-            self.push_inner(ComputeCommand::CancelPeeks {
+            metric.set(1);
+            self.reduced_count += live_cancels.len();
```

Comment:
As you mentioned in a comment, the metric is increased by only one by the command count in `reduced_count` by the length of `live_cancels`. It would be ideal if we would make this consistent. At least originally, the command history size metric was intended to count the physical number of commands in the history, while other logical metrics (like dataflow count or in the future live dataflow count or live peek count) could be used to complement it. That way, we get a a picture of how the representation looks like.